lib/Dialect/Handshake/HandshakeOps.cpp

```diff
@@ -134,9 +145,46 @@ void ForkOp::build(OpBuilder &builder, OperationState &result, Value operand,
   result.addAttribute("control", builder.getBoolAttr(isControl));
 }
 
+namespace {
+
+/// A pattern to remove unused fork results. This pattern will not run on
+/// user-provided IR (since unused results are illegal in parsed IR). However,
+/// the pattern is applicable during canonicalization where we might remove
+/// other operations that use the fork result.
+struct EliminateUnusedForkResultsPattern : mlir::OpRewritePattern<ForkOp> {
+  using mlir::OpRewritePattern<ForkOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(ForkOp op,
+                                PatternRewriter &rewriter) const override {
+    std::set<unsigned> unusedIndexes;
+
+    for (auto res : llvm::enumerate(op.getResults()))
+      if (res.value().getUses().empty())
+        unusedIndexes.insert(res.index());
+
+    if (unusedIndexes.size() == 0)
+      return failure();
+
+    // Create a new fork op, dropping the unused results.
+    rewriter.setInsertionPoint(op);
+    auto newFork =
+        rewriter.create<ForkOp>(op.getLoc(), op.getOperand(),
+                                op.getNumResults() - unusedIndexes.size());
+    unsigned i = 0;
+    for (auto oldRes : llvm::enumerate(op.getResults()))
+      if (unusedIndexes.count(oldRes.index()) == 0)
+        oldRes.value().replaceAllUsesWith(newFork.getResult(i++));
```

Comment:
Please use the `rewriter` rather than using the `value().replaceAllUsesWith()` API directly. I think there is a method like `rewriter.replaceOpWithNewOp`. If that won't work since you are changing the number of results, I think you must use `rewriter.updateRootInPlace` for this to be correct.

Reply:
`replaceOpWithNewOp` is indeed not used is due to the constraint on # of result of the replaced op. 
The rewriter will fail if we `rewriter.eraseOp(op);` and `op` still has used (happened during implementing this). I'm not sure i see how `updateRootInPlace` would add/change anything in the situation.

Reply:
I don't think it's legal to say

```
  oldRes.value().replaceAllUsesWith(newFork.getResult(i++));
```

I think you instead have to say

```
rewriter.updateRootInPlace(op, [&](){
  oldRes.value().replaceAllUsesWith(newFork.getResult(i++));
});
```

It might not manifest on the current test cases, but in general all IR mutation must happen through the PatternRewriter: https://mlir.llvm.org/docs/PatternRewriter/#restrictions